pkg/model/db-statefulset.go

```diff
@@ -0,0 +1,108 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package model
+
+import (
+	"fmt"
+	"os"
+
+	corev1 "k8s.io/api/core/v1"
+
+	bsv1alpha1 "janus-idp.io/backstage-operator/api/v1alpha1"
+	"janus-idp.io/backstage-operator/pkg/utils"
+	appsv1 "k8s.io/api/apps/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const PostgresImageEnvVar = "RELATED_IMAGE_postgresql"
+
+type DbStatefulSetFactory struct{}
+
+func (f DbStatefulSetFactory) newBackstageObject() BackstageObject {
+	return &DbStatefulSet{statefulSet: &appsv1.StatefulSet{}}
+}
+
+type DbStatefulSet struct {
+	statefulSet *appsv1.StatefulSet
+	secretName  string
+}
+
+func init() {
+	registerConfig("db-statefulset.yaml", DbStatefulSetFactory{}, ForLocalDatabase)
+}
+
+// implementation of BackstageObject interface
+func (b *DbStatefulSet) Object() client.Object {
+	// override image with env var
+	if os.Getenv(PostgresImageEnvVar) != "" {
+		b.container().Image = os.Getenv(PostgresImageEnvVar)
+	}
+	return b.statefulSet
+}
+
+// implementation of BackstageObject interface
+func (b *DbStatefulSet) initMetainfo(backstageMeta bsv1alpha1.Backstage, ownsRuntime bool) {
+	initMetainfo(b, backstageMeta, ownsRuntime)
+	b.statefulSet.SetName(utils.GenerateRuntimeObjectName(backstageMeta.Name, "db-statefulset"))
+	utils.GenerateLabel(&b.statefulSet.Spec.Template.ObjectMeta.Labels, backstageAppLabel, fmt.Sprintf("backstage-db-%s", backstageMeta.Name))
+	utils.GenerateLabel(&b.statefulSet.Spec.Selector.MatchLabels, backstageAppLabel, fmt.Sprintf("backstage-db-%s", backstageMeta.Name))
+}
+
+// implementation of BackstageObject interface
+func (b *DbStatefulSet) addToModel(model *RuntimeModel) {
+	model.localDbStatefulSet = b
+}
+
+// implementation of BackstageObject interface
+func (b *DbStatefulSet) EmptyObject() client.Object {
+	return &appsv1.StatefulSet{}
+}
+
+// implementation of BackstageObject interface
+func (b *DbStatefulSet) validate(model *RuntimeModel) error {
+	return nil
+}
+
+// Injects DB Secret name as an env variable of DB container
+// Local DB pod considered to have single container
+func (b *DbStatefulSet) setSecretNameEnvFrom(envFrom corev1.EnvFromSource) {
+
+	// it is possible that Secret name already set by default configuration
+	// has to be overriden in this case
+	if b.secretName != "" {
+		var ind int
+		for i, v := range b.container().EnvFrom {
+			if v.SecretRef.Name == b.secretName {
+				ind = i
+				break
+			}
+		}
+		b.statefulSet.Spec.Template.Spec.Containers[0].EnvFrom[ind] = envFrom
+
+	} else {
+		b.statefulSet.Spec.Template.Spec.Containers[0].EnvFrom = append(b.statefulSet.Spec.Template.Spec.Containers[0].EnvFrom, envFrom)
+	}
+	b.secretName = envFrom.SecretRef.Name
+}
+
+// returns DB container
+func (b *DbStatefulSet) container() *corev1.Container {
+	return &b.podSpec().Containers[0]
```

Comment:
It seems this list of containers might be empty. I tried to apply the [`examples/janus-cr-with-app-configs.yaml` CF](https://github.com/janus-idp/operator/blob/main/examples/janus-cr-with-app-configs.yaml) (after renaming `spec.enableLocalDb` into `spec.database.enabledLocalDb` to fix the CR validation error), and the controller keeps crashing while doing the reconciliation:

```
2024-01-10T16:19:19Z    INFO    Observed a panic in reconciler: runtime error: index out of range [0] with length 0     {"controller": "backstage", "controllerGroup": "janus-idp.io", "control
lerKind": "Backstage", "Backstage": {"name":"my-backstage-app-with-app-config","namespace":"my-ns"}, "namespace": "my-ns", "name": "my-backstage-app-with-app-config", "reconcileID": "76864ec0
-6758-4d23-93a2-c5036c133d7f"}
panic: runtime error: index out of range [0] with length 0 [recovered]
        panic: runtime error: index out of range [0] with length 0

goroutine 163 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:115 +0x1fa
panic({0x18125a0, 0xc000fcc690})
        /usr/lib/golang/src/runtime/panic.go:884 +0x213
janus-idp.io/backstage-operator/pkg/model.(*DbStatefulSet).container(...)
        /workspace/pkg/model/db-statefulset.go:102
janus-idp.io/backstage-operator/pkg/model.(*DbStatefulSet).Object(0xc0014a00c0)
        /workspace/pkg/model/db-statefulset.go:50 +0xe6
janus-idp.io/backstage-operator/pkg/model.InitObjects({_, _}, {{{0x15ecb9a, 0x9}, {0xc00138c180, 0x15}}, {{0xc0000d6b20, 0x20}, {0x0, 0x0}, ...}, ...}, ...)
        /workspace/pkg/model/runtime.go:94 +0x1c4
janus-idp.io/backstage-operator/controllers.(*BackstageReconciler).Reconcile(0xc000539980, {0x1b7b180, 0xc000391cb0}, {{{0xc0004f42d6, 0x5}, {0xc0000d6b20, 0x20}}})
        /workspace/controllers/backstage_controller.go:109 +0x3f1
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x1b7b180?, {0x1b7b180?, 0xc000391cb0?}, {{{0xc0004f42d6?, 0x16a1600?}, {0xc0000d6b20?, 0x1b68bd8?}}})
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:118 +0xc8
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc000468320, {0x1b7b0d8, 0xc0001186e0}, {0x1798100?, 0xc0002fd220?})
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:314 +0x377
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc000468320, {0x1b7b0d8, 0xc0001186e0})
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:265 +0x1d9
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:226 +0x85
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:222 +0x587
```

<details>
<summary>Full logs</summary>

```
$ kubectl logs -n openshift-operators backstage-controller-manager-6b9f4bc857-vc89c                                                                                                            
Alias tip: kl -n openshift-operators backstage-controller-manager-6b9f4bc857-vc89c                                                                                                             
2024-01-10T16:18:51Z    INFO    controller-runtime.metrics      Metrics server is starting to listen    {"addr": "127.0.0.1:8080"}                                                             
2024-01-10T16:18:51Z    INFO    setup   starting manager with parameters:       {"own-runtime": true, "env.LOCALBIN": "", "isOpenShift": true}                                                 
2024-01-10T16:18:51Z    INFO    starting server {"path": "/metrics", "kind": "metrics", "addr": "127.0.0.1:8080"}                                                                              
2024-01-10T16:18:51Z    INFO    Starting server {"kind": "health probe", "addr": "[::]:8081"}                                                                                                  
I0110 16:18:51.338651       1 leaderelection.go:250] attempting to acquire leader lease openshift-operators/06bdbdd5.janus-idp.io...                                                           
I0110 16:19:08.268597       1 leaderelection.go:260] successfully acquired lease openshift-operators/06bdbdd5.janus-idp.io                                                                     
2024-01-10T16:19:08Z    DEBUG   events  backstage-controller-manager-6b9f4bc857-vc89c_a06f486e-5aa0-4493-a2c7-c4711d31ae2a became leader        {"type": "Normal", "object": {"kind":"Lease","n
amespace":"openshift-operators","name":"06bdbdd5.janus-idp.io","uid":"30dea8a2-061c-4899-a575-87d44c191ebf","apiVersion":"coordination.k8s.io/v1","resourceVersion":"11933"}, "reason": "Leader
Election"}                                                                                                                                                                                     
2024-01-10T16:19:08Z    INFO    Starting EventSource    {"controller": "backstage", "controllerGroup": "janus-idp.io", "controllerKind": "Backstage", "source": "kind source: *v1alpha1.Backsta
ge"}                                                                                                                                                                                           
2024-01-10T16:19:08Z    INFO    Starting EventSource    {"controller": "backstage", "controllerGroup": "janus-idp.io", "controllerKind": "Backstage", "source": "kind source: *v1.Deployment"} 
2024-01-10T16:19:08Z    INFO    Starting EventSource    {"controller": "backstage", "controllerGroup": "janus-idp.io", "controllerKind": "Backstage", "source": "kind source: *v1.Service"}    
2024-01-10T16:19:08Z    INFO    Starting EventSource    {"controller": "backstage", "controllerGroup": "janus-idp.io", "controllerKind": "Backstage", "source": "kind source: *v1.StatefulSet"}
2024-01-10T16:19:08Z    INFO    Starting Controller     {"controller": "backstage", "controllerGroup": "janus-idp.io", "controllerKind": "Backstage"}                                          
2024-01-10T16:19:08Z    INFO    Starting workers        {"controller": "backstage", "controllerGroup": "janus-idp.io", "controllerKind": "Backstage", "worker count": 1}                       
2024-01-10T16:19:14Z    DEBUG   starting reconciliation (namespace: "my-ns/my-backstage-app-with-app-config")   {"controller": "backstage", "controllerGroup": "janus-idp.io", "controllerKind"
: "Backstage", "Backstage": {"name":"my-backstage-app-with-app-config","namespace":"my-ns"}, "namespace": "my-ns", "name": "my-backstage-app-with-app-config", "reconcileID": "d592d4ea-c0ef-40
79-9d53-80c2b86f5e4a"}                                                                                                                                                                         
2024-01-10T16:19:14Z    ERROR   Reconciler error        {"controller": "backstage", "controllerGroup": "janus-idp.io", "controllerKind": "Backstage", "Backstage": {"name":"my-backstage-app-wi
th-app-config","namespace":"my-ns"}, "namespace": "my-ns", "name": "my-backstage-app-with-app-config", "reconcileID": "d592d4ea-c0ef-4079-9d53-80c2b86f5e4a", "error": "failed to preprocess ba
ckstage spec: failed to get configMap my-backstage-config-backend-auth: ConfigMap \"my-backstage-config-backend-auth\" not found"}                                                             
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler                                                                                                          
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:324
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:265
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:226
2024-01-10T16:19:14Z    DEBUG   starting reconciliation (namespace: "my-ns/my-backstage-app-with-app-config")   {"controller": "backstage", "controllerGroup": "janus-idp.io", "controllerKind"
: "Backstage", "Backstage": {"name":"my-backstage-app-with-app-config","namespace":"my-ns"}, "namespace": "my-ns", "name": "my-backstage-app-with-app-config", "reconcileID": "60a9410a-a75c-4f
3c-9d6c-93b704afe493"}
2024-01-10T16:19:14Z    ERROR   Reconciler error        {"controller": "backstage", "controllerGroup": "janus-idp.io", "controllerKind": "Backstage", "Backstage": {"name":"my-backstage-app-wi
th-app-config","namespace":"my-ns"}, "namespace": "my-ns", "name": "my-backstage-app-with-app-config", "reconcileID": "60a9410a-a75c-4f3c-9d6c-93b704afe493", "error": "failed to preprocess ba
ckstage spec: failed to get configMap my-backstage-config-backend-auth: ConfigMap \"my-backstage-config-backend-auth\" not found"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:324
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:265
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:226
2024-01-10T16:19:14Z    DEBUG   starting reconciliation (namespace: "my-ns/my-backstage-app-with-app-config")   {"controller": "backstage", "controllerGroup": "janus-idp.io", "controllerKind"
: "Backstage", "Backstage": {"name":"my-backstage-app-with-app-config","namespace":"my-ns"}, "namespace": "my-ns", "name": "my-backstage-app-with-app-config", "reconcileID": "80fe151b-137d-4a
5f-b684-7572d158ce82"}
2024-01-10T16:19:14Z    ERROR   Reconciler error        {"controller": "backstage", "controllerGroup": "janus-idp.io", "controllerKind": "Backstage", "Backstage": {"name":"my-backstage-app-wi
th-app-config","namespace":"my-ns"}, "namespace": "my-ns", "name": "my-backstage-app-with-app-config", "reconcileID": "80fe151b-137d-4a5f-b684-7572d158ce82", "error": "failed to preprocess ba
ckstage spec: failed to get configMap my-backstage-config-backend-auth: ConfigMap \"my-backstage-config-backend-auth\" not found"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:324
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:265
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:226
2024-01-10T16:19:14Z    DEBUG   starting reconciliation (namespace: "my-ns/my-backstage-app-with-app-config")   {"controller": "backstage", "controllerGroup": "janus-idp.io", "controllerKind"
: "Backstage", "Backstage": {"name":"my-backstage-app-with-app-config","namespace":"my-ns"}, "namespace": "my-ns", "name": "my-backstage-app-with-app-config", "reconcileID": "15e036c6-c2e5-4f
0b-8c4d-083ba0aca9ed"}
2024-01-10T16:19:14Z    ERROR   Reconciler error        {"controller": "backstage", "controllerGroup": "janus-idp.io", "controllerKind": "Backstage", "Backstage": {"name":"my-backstage-app-wi
th-app-config","namespace":"my-ns"}, "namespace": "my-ns", "name": "my-backstage-app-with-app-config", "reconcileID": "15e036c6-c2e5-4f0b-8c4d-083ba0aca9ed", "error": "failed to preprocess ba
ckstage spec: failed to get configMap my-backstage-config-backend-auth: ConfigMap \"my-backstage-config-backend-auth\" not found"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:324
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:265
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2                                                                                          17:19:34 [78/55999]
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:226
2024-01-10T16:19:14Z    DEBUG   starting reconciliation (namespace: "my-ns/my-backstage-app-with-app-config")   {"controller": "backstage", "controllerGroup": "janus-idp.io", "controllerKind"
: "Backstage", "Backstage": {"name":"my-backstage-app-with-app-config","namespace":"my-ns"}, "namespace": "my-ns", "name": "my-backstage-app-with-app-config", "reconcileID": "bedb5b6c-3206-46
c8-bc07-9674d54791b0"}
2024-01-10T16:19:14Z    ERROR   Reconciler error        {"controller": "backstage", "controllerGroup": "janus-idp.io", "controllerKind": "Backstage", "Backstage": {"name":"my-backstage-app-wi
th-app-config","namespace":"my-ns"}, "namespace": "my-ns", "name": "my-backstage-app-with-app-config", "reconcileID": "bedb5b6c-3206-46c8-bc07-9674d54791b0", "error": "failed to preprocess ba
ckstage spec: failed to get configMap my-backstage-config-backend-auth: ConfigMap \"my-backstage-config-backend-auth\" not found"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:324
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:265
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:226
2024-01-10T16:19:14Z    DEBUG   starting reconciliation (namespace: "my-ns/my-backstage-app-with-app-config")   {"controller": "backstage", "controllerGroup": "janus-idp.io", "controllerKind"
: "Backstage", "Backstage": {"name":"my-backstage-app-with-app-config","namespace":"my-ns"}, "namespace": "my-ns", "name": "my-backstage-app-with-app-config", "reconcileID": "94bfb2a0-d7b9-44
a6-949d-12956fc814fc"}
2024-01-10T16:19:14Z    ERROR   Reconciler error        {"controller": "backstage", "controllerGroup": "janus-idp.io", "controllerKind": "Backstage", "Backstage": {"name":"my-backstage-app-wi
th-app-config","namespace":"my-ns"}, "namespace": "my-ns", "name": "my-backstage-app-with-app-config", "reconcileID": "94bfb2a0-d7b9-44a6-949d-12956fc814fc", "error": "failed to preprocess ba
ckstage spec: failed to get configMap my-backstage-config-backend-auth: ConfigMap \"my-backstage-config-backend-auth\" not found"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:324
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:265
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:226
2024-01-10T16:19:15Z    DEBUG   starting reconciliation (namespace: "my-ns/my-backstage-app-with-app-config")   {"controller": "backstage", "controllerGroup": "janus-idp.io", "controllerKind"
: "Backstage", "Backstage": {"name":"my-backstage-app-with-app-config","namespace":"my-ns"}, "namespace": "my-ns", "name": "my-backstage-app-with-app-config", "reconcileID": "1879853c-5152-4e
24-85be-9349a2f025b5"}
2024-01-10T16:19:15Z    ERROR   Reconciler error        {"controller": "backstage", "controllerGroup": "janus-idp.io", "controllerKind": "Backstage", "Backstage": {"name":"my-backstage-app-wi
th-app-config","namespace":"my-ns"}, "namespace": "my-ns", "name": "my-backstage-app-with-app-config", "reconcileID": "1879853c-5152-4e24-85be-9349a2f025b5", "error": "failed to preprocess ba
ckstage spec: failed to get configMap my-backstage-config-cm1: ConfigMap \"my-backstage-config-cm1\" not found"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:324
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:265
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:226
2024-01-10T16:19:15Z    DEBUG   starting reconciliation (namespace: "my-ns/my-backstage-app-with-app-config")   {"controller": "backstage", "controllerGroup": "janus-idp.io", "controllerKind"
: "Backstage", "Backstage": {"name":"my-backstage-app-with-app-config","namespace":"my-ns"}, "namespace": "my-ns", "name": "my-backstage-app-with-app-config", "reconcileID": "5de86128-5e59-4f
88-af1b-b47a4a58ffe8"}
2024-01-10T16:19:15Z    ERROR   Reconciler error        {"controller": "backstage", "controllerGroup": "janus-idp.io", "controllerKind": "Backstage", "Backstage": {"name":"my-backstage-app-wi
th-app-config","namespace":"my-ns"}, "namespace": "my-ns", "name": "my-backstage-app-with-app-config", "reconcileID": "5de86128-5e59-4f88-af1b-b47a4a58ffe8", "error": "failed to preprocess ba
ckstage spec: failed to get configMap my-backstage-config-cm1: ConfigMap \"my-backstage-config-cm1\" not found"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:324
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:265
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:226
2024-01-10T16:19:16Z    DEBUG   starting reconciliation (namespace: "my-ns/my-backstage-app-with-app-config")   {"controller": "backstage", "controllerGroup": "janus-idp.io", "controllerKind"
: "Backstage", "Backstage": {"name":"my-backstage-app-with-app-config","namespace":"my-ns"}, "namespace": "my-ns", "name": "my-backstage-app-with-app-config", "reconcileID": "15f2ccc4-dccb-4f
e2-a121-5f09af935046"}
2024-01-10T16:19:16Z    ERROR   Reconciler error        {"controller": "backstage", "controllerGroup": "janus-idp.io", "controllerKind": "Backstage", "Backstage": {"name":"my-backstage-app-wi
th-app-config","namespace":"my-ns"}, "namespace": "my-ns", "name": "my-backstage-app-with-app-config", "reconcileID": "15f2ccc4-dccb-4fe2-a121-5f09af935046", "error": "failed to preprocess ba
ckstage spec: failed to get ConfigMap my-backstage-extra-files-cm1: ConfigMap \"my-backstage-extra-files-cm1\" not found"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:324
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:265
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:226
2024-01-10T16:19:17Z    DEBUG   starting reconciliation (namespace: "my-ns/my-backstage-app-with-app-config")   {"controller": "backstage", "controllerGroup": "janus-idp.io", "controllerKind"
: "Backstage", "Backstage": {"name":"my-backstage-app-with-app-config","namespace":"my-ns"}, "namespace": "my-ns", "name": "my-backstage-app-with-app-config", "reconcileID": "441237ee-15e3-41
bf-90f8-47c0577a575e"}
2024-01-10T16:19:17Z    ERROR   Reconciler error        {"controller": "backstage", "controllerGroup": "janus-idp.io", "controllerKind": "Backstage", "Backstage": {"name":"my-backstage-app-wi
th-app-config","namespace":"my-ns"}, "namespace": "my-ns", "name": "my-backstage-app-with-app-config", "reconcileID": "441237ee-15e3-41bf-90f8-47c0577a575e", "error": "failed to preprocess ba
ckstage spec: failed to get ConfigMap my-backstage-extra-files-cm1: ConfigMap \"my-backstage-extra-files-cm1\" not found"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:324
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:265
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:226
2024-01-10T16:19:19Z    DEBUG   starting reconciliation (namespace: "my-ns/my-backstage-app-with-app-config")   {"controller": "backstage", "controllerGroup": "janus-idp.io"17:19:34 [4/55999]
: "Backstage", "Backstage": {"name":"my-backstage-app-with-app-config","namespace":"my-ns"}, "namespace": "my-ns", "name": "my-backstage-app-with-app-config", "reconcileID": "76864ec0-6758-4d
23-93a2-c5036c133d7f"}
2024-01-10T16:19:19Z    DEBUG   failed to read default value for optional key. Ignored 
        {"controller": "backstage", "controllerGroup": "janus-idp.io", "controllerKind": "Backstage", "Backstage": {"name":"my-backstage-app-with-app-config","namespace":"my-ns"}, "namespace"
: "my-ns", "name": "my-backstage-app-with-app-config", "reconcileID": "76864ec0-6758-4d23-93a2-c5036c133d7f", "configmap-envs.yaml": "failed to read default value for the key configmap-envs.y
aml, reason: failed to read YAML file: open default-config/configmap-envs.yaml: no such file or directory"}
2024-01-10T16:19:19Z    DEBUG   failed to read default value for optional key. Ignored 
        {"controller": "backstage", "controllerGroup": "janus-idp.io", "controllerKind": "Backstage", "Backstage": {"name":"my-backstage-app-with-app-config","namespace":"my-ns"}, "namespace"
: "my-ns", "name": "my-backstage-app-with-app-config", "reconcileID": "76864ec0-6758-4d23-93a2-c5036c133d7f", "configmap-files.yaml": "failed to read default value for the key configmap-files
.yaml, reason: failed to read YAML file: open default-config/configmap-files.yaml: no such file or directory"}
2024-01-10T16:19:19Z    INFO    Observed a panic in reconciler: runtime error: index out of range [0] with length 0     {"controller": "backstage", "controllerGroup": "janus-idp.io", "control
lerKind": "Backstage", "Backstage": {"name":"my-backstage-app-with-app-config","namespace":"my-ns"}, "namespace": "my-ns", "name": "my-backstage-app-with-app-config", "reconcileID": "76864ec0
-6758-4d23-93a2-c5036c133d7f"}
panic: runtime error: index out of range [0] with length 0 [recovered]
        panic: runtime error: index out of range [0] with length 0

goroutine 163 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:115 +0x1fa
panic({0x18125a0, 0xc000fcc690})
        /usr/lib/golang/src/runtime/panic.go:884 +0x213
janus-idp.io/backstage-operator/pkg/model.(*DbStatefulSet).container(...)
        /workspace/pkg/model/db-statefulset.go:102
janus-idp.io/backstage-operator/pkg/model.(*DbStatefulSet).Object(0xc0014a00c0)
        /workspace/pkg/model/db-statefulset.go:50 +0xe6
janus-idp.io/backstage-operator/pkg/model.InitObjects({_, _}, {{{0x15ecb9a, 0x9}, {0xc00138c180, 0x15}}, {{0xc0000d6b20, 0x20}, {0x0, 0x0}, ...}, ...}, ...)
        /workspace/pkg/model/runtime.go:94 +0x1c4
janus-idp.io/backstage-operator/controllers.(*BackstageReconciler).Reconcile(0xc000539980, {0x1b7b180, 0xc000391cb0}, {{{0xc0004f42d6, 0x5}, {0xc0000d6b20, 0x20}}})
        /workspace/controllers/backstage_controller.go:109 +0x3f1
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x1b7b180?, {0x1b7b180?, 0xc000391cb0?}, {{{0xc0004f42d6?, 0x16a1600?}, {0xc0000d6b20?, 0x1b68bd8?}}})
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:118 +0xc8
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc000468320, {0x1b7b0d8, 0xc0001186e0}, {0x1798100?, 0xc0002fd220?})
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:314 +0x377
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc000468320, {0x1b7b0d8, 0xc0001186e0})
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:265 +0x1d9
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:226 +0x85
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:222 +0x587
```

</details>

Reply:
Hmm, actually it also happened with the minimal `bs1.yaml` CR: https://github.com/janus-idp/operator/blob/main/examples/bs1.yaml
Am I missing something?

Reply:
To clarify, this happened after building the container images and running the controller in the cluster (`make release-build release-push catalog-update`).
Now when running locally with `make install run`, I'm seeing a different error: it seems `<RELATED_IMAGE_backstage>` and `<RELATED_IMAGE_postgresql>` are not being replaced (I didn't set those env vars before running the operator):
```
$ kubectl describe pod bs1-db-statefulset-0
[...]
Events:
  Type     Reason                  Age                     From                     Message
  ----     ------                  ----                    ----                     -------
  Normal   Scheduled               5m40s                   default-scheduler        Successfully assigned my-ns/bs1-db-statefulset-0 to ci-ln-ywvkv12-72292-56c76-worker-b-lb5r2
  Normal   SuccessfulAttachVolume  5m36s                   attachdetach-controller  AttachVolume.Attach succeeded for volume "pvc-07468e13-d1c3-456e-8a9d-800ee7ec1bf1"
  Normal   AddedInterface          5m32s                   multus                   Add eth0 [10.128.2.16/23] from ovn-kubernetes
  Warning  Failed                  3m33s (x12 over 5m32s)  kubelet                  Error: InvalidImageName
  Warning  InspectFailed           31s (x26 over 5m32s)    kubelet                  Failed to apply default image tag "<RELATED_IMAGE_postgresql>": couldn't parse image reference "<RELATED_IMAGE_postgresql>": invalid reference format: repository name must be lowercase
```

Reply:
is not it gone after last fixes?

Reply:
Nope, I am still seeing a panic. It seems to happen only when `RELATED_IMAGE_*` env var are set.
You can reproduce locally with something like:
```sh
$ RELATED_IMAGE_backstage=busybox make install run
```

Then you'll see the panic when reconciling any CR.

Reply:
It should be OK now

Running as you suggested:
Error: failed to start container "backstage-backend": Error response from daemon: failed to create shim task: OCI runtime create failed: runc  │
│ create failed: unable to start container process: exec: "--config": executable file not found in $PATH: unknown 

But it is expected I think as the image is just busybox

Reply:
Nice, with the latest changes, I'm no longer facing a panic.
But I noticed one more thing to change on the Backstage Pod.
The image should be replaced on **all** containers of the Pod Spec, including the init container. Currently, it is only replaced for the main container. Init container is still using the default `quay.io/janus-idp/backstage-showcase:next` image, which might be problematic for downstream builds or disconnected environments.
Example if I set `RELATED_IMAGE_backstage` to `quay.io/rhdh/rhdh-hub-rhel9:1.0-200` in the controller environment:
```
$ kubectl get deployment bs1-deployment -o yaml
[...]
spec:
 template:
    spec:
      containers:
        image: quay.io/rhdh/rhdh-hub-rhel9:1.0-200
        imagePullPolicy: IfNotPresent
        name: backstage-backend
      initContainers:
        image: quay.io/janus-idp/backstage-showcase:next
        imagePullPolicy: IfNotPresent
        name: install-dynamic-plugins
[...]
```

Reply:
Same behavior when overriding the image from the one set in `spec.application.image`. It should override the image on **all** containers of the Pod Spec, including the init container. Currently, it only overrides the main container image.

Reply:
FYI, we had a `visitContainers` utility function to consistently perform some operations on a `PodTemplateSpec`: https://github.com/janus-idp/operator/blob/bac627410197f4b1d6a120b55189d7464181fa3c/controllers/backstage_deployment.go#L136-L147

I think it can help prevent such issues in the future.